op-e2e/actions/interop/proofs_test.go

```diff
@@ -1391,15 +1397,61 @@ func WithInteropEnabled(t helpers.StatefulTesting, actors *dsl.InteropActors, de
 		f.DependencySet = depSet
 
 		for _, chain := range []*dsl.Chain{actors.ChainA, actors.ChainB} {
+			verifier, canonicalOnlyEngine := createVerifierWithOnlyCanonicalBlocks(t, actors.L1Miner, chain)
 			f.L2Sources = append(f.L2Sources, &fpHelpers.FaultProofProgramL2Source{
-				Node:        chain.Sequencer.L2Verifier,
-				Engine:      chain.SequencerEngine,
+				Node:        verifier,
+				Engine:      canonicalOnlyEngine,
 				ChainConfig: chain.L2Genesis.Config,
 			})
 		}
 	}
 }
 
+// createVerifierWithOnlyCanonicalBlocks creates a new L2Verifier and associated L2Engine that only has the canonical
+// blocks from chain in its database. Non-canonical blocks, their world state, receipts and other data are not available
+func createVerifierWithOnlyCanonicalBlocks(t helpers.StatefulTesting, l1Miner *helpers.L1Miner, chain *dsl.Chain) (*helpers.L2Verifier, *helpers.L2Engine) {
```

Comment:
This is a pretty slick. Alternatively, it should suffice to wrap the original verifier node with a filter that returns `NotFound` for non-canon blocks. Though I like this method for its thoroughness; which covers the off-chance that a program execution needs some preimage without also needing the block header it's included in.

Reply:
Yeah I really wanted to have the test confirm that we actually stored all the required preimages. It would be easy for the re-generation to store the block and transactions but not store all the receipts for example. Those would still be in the database if we just filtered out the block hash and it would be really hard to accurately filter out all the details of a block (but not things from canonical blocks that have the same hash which happens a lot).